### DISCUSSION TITLE
gcc Werror about closing_iterator: -Werror=strict-overflow

### DISCUSSION THREAD

**Discussion by wxy-creat** [boostorg/geometry#1339] [Q&A] [answered]
Hello, I am getting: 

` warning: assuming signed overflow does not occur when assuming that (X + c) < X is always false [-Wstrict-overflow]`

on this line in file: closing_iterator.hpp: 148:
`if (m_index-- < m_size)`

the complete codes are:
`    inline void decrement()
    {
        if (m_index-- < m_size)
        {
            --m_iterator;
        }
        else
        {
            update_iterator();
        }
    }`

After checking other codes in this header file, I guess this warning might  be caused by  the constructor of struct closing_iterator:
`m_index((m_size == 0) ? 0 : m_size + 1)`
In this case, if `m_index = m_size + 1` meanwhile `m_size` is already the maximum signed value, m_index would be overflow causing undefined behavior.
I know I can use `GCC diagnostic ignored "-Wstrict-overflow"` or any other way to ignore this warning, but it would be great to know if there is any evidence to make sure that this overflow will never happen.
Thanks a lot.

**Comment by tinko92** [comment, marked as answer]
I cannot replicate it, do you have a full example with GCC version and command line that generates this warning? I can only try to make a guess: If I understand this class and the warning (https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#:~:text=It%20warns%20about%20cases,on%20the%20optimization%20level.) correctly then this would mean, GCC makes some optimization that is only valid if no integer overflow occurs.

Maybe something like this: GCC sees that this class is instantiated with the end constructor and then sees several decrement-calls occurring in a loop. It deduces that this condition `m_index-- < m_size` will only ever be false in the first call to decrement, and in all subsequent calls, it must be true, so for the remainder of the loop it can generate the code as if it only said `--m_index;--m_iterator;`. If that was what is going on, then I think the warning would make sense to suppress because if m_size + 1 overflows or if decrement is called so many times that m_index overflows, I think, this facility should break in other ways as well.

**Comment by wxy-creat** [comment]
Thanks a lot. I think you are right. My GCC version is gcc-12.3.0 and I use -O2 to open compiler optimization. It can work without -O2.
